Feat/deploy and fund

pappas999 opened the pull request:
### Auto Funding Feature Implementation

**Goal**

- Implement configurable functionality for Auto-funding of contracts that require LINK (Any-API and VRF). The idea behind this is that every time a developer deploys these contracts, they must be funded with LINK before they can make an oracle request. Sometimes they may do dozens of contract deployments in a row as they debug issues, so having the funding of the contract done automatically during the deployment can save them time and effort.
- Make the amount of LINK sent to deployed contracts configurable 

**Implementation**

- Two new config settings have been introduced per-network 'autoFund' and 'fundAmount', which are the main drivers as to whether the Starter Kit should attempt an auto-fund on contract deployment, and if so, how much to fund the contract with. The new fundAmount config is used for manual funding of contracts as well, making it fully configurable if the developer requires it.
- The default config is to always try and auto-fund on non-mainnet networks. The deployment scripts will first check to see if the auto-funding config is enabled for the network, then they will check to see if the deploying address has enough LINK to satisfy the auto-funding or not. If it does, it will auto-fund the contract once it's deployed, otherwise, it will raise a warning to the user that it couldn't auto-fund due to insufficient LINK. This way, developers without any LINK in their deploying account won't get any errors when deploying the contracts.

alexroan commented on tasks/fund-link.js:
```diff
@@ -1,4 +1,3 @@
-
```
Could we add a task to `withdrawLink` also?

pappas999 replied:
great suggestion! I've added it in.

alexroan commented on tasks/fund-link.js:
```diff
@@ -7,21 +6,22 @@ task("fund-link", "Funds a contract with LINK")
         const accounts = await ethers.getSigners()
         const signer = accounts[0]
-
+        
```
```suggestion

```

alexroan commented on helper-hardhat-config.js:
```diff
@@ -30,14 +36,20 @@ const networkConfig = {
         name: 'goerli',
-        linkToken: '0x326c977e6efc84e512bb9c30f76e30c160ed06fb'
+        linkToken: '0x326c977e6efc84e512bb9c30f76e30c160ed06fb',
+        autoFund: '0',
```
How come this is `0` and the other testnets are `1000000000000000000`?

pappas999 replied:
Looking at the [docs](https://docs.chain.link/docs/vrf-contracts) for VRF & Any-API, don't think they are live on Goerli testnet. So there's no need to pre-fund by default

alexroan reviewed (changes requested):
Nice addition. A couple of comments from brushing over the code.

PatrickAlphaC commented on deploy/02_Deploy_APIConsumer.js:
```diff
@@ -23,17 +25,47 @@ module.exports = async ({
+    const LinkToken = await ethers.getContractFactory("LinkToken")
+    const linkTokenContract = new ethers.Contract(linkTokenAddress, LinkToken.interface, signer)
+    var balanceHex = await linkTokenContract.balanceOf(signer.address);
+    var balance = await web3.utils.toBN(balanceHex._hex).toString(); 
```
use let over `var`

pappas999 commented:
- added withdraw link task. Checks to see if the contract has any LINK first
- Changed 'var' to 'let' 
@PatrickAlphaC @alexroan

pappas999 commented:
modified autoFund param to boolean at request of @PatrickAlphaC

PatrickAlphaC reviewed (changes requested):
Can we modularize the auto-funding? We can put it in the helper config. Right now we have a bunch of code in the VRF and API deployment that does the same thing.

PatrickAlphaC commented:
Can we modularize the auto-funding? Let me know if this makes sense.

pappas999 commented:
modularized the autoFund as requested. Also made some changes to the tests so that they work on public networks as well as local ones.

Basically, the testing is split into 2 scenarios, 1 is local and 2 is public networks. If it's local, it uses the existing tests that were there, if it's a public network, it doesn't use any mocks or aggregators etc, and does real tests. Ie for any-api and vrf ,it calls the function to do the request, then waits 30 secs for the oracle to callback, then checks the result.  Tested both against local and kovan network

@PatrickAlphaC

PatrickAlphaC commented on helper-hardhat-config.js:
```diff
@@ -50,8 +62,45 @@ const getNetworkIdFromName = async (networkIdName) => {
     return null
 }
 
+const autoFundCheck = async (contractAddr, networkName, linkTokenAddress, additionalMessage) => {
```
The fund check shouldn't also do the funding. The fund check should just check if it can be funded. If yes, then it should run `autoFund`

PatrickAlphaC commented on test/APIConsumer_test.js:
```diff
@@ -1,29 +1,55 @@
+      const tx_receipt = await transaction.wait()
+      const requestId = tx_receipt.events[0].topics[1]
+      const returnData = web3.utils.padLeft(web3.utils.padLeft(web3.utils.toHex(expected)), 64)
+      const tx = await mockOracle.fulfillOracleRequest(requestId, returnData)
```
could we just have the if here?

ie: if local -> call fulfilment. Right now, we basically have to write two totally different tests for local and testnet, when its really only 1 function that we need to call.

pappas999 replied:
have gotten rid of the if/else, and just have if as requested. The only problem is, once again you can't have 2 different tests in this case, you can't compare the local result to the expected variable, but then for the public network check that its > 0. So to make it more uniform, both scenarios now just check for > 0 result

PatrickAlphaC commented on test/PriceConsumerV3_test.js:
```diff
@@ -4,20 +4,31 @@ const { networkConfig } = require('../helper-hardhat-config')
   it('should return a positive value', async () => {
-    expect(await priceConsumerV3.getLatestPrice())
-      .to.equal((await ethUsdAggregator.latestRoundData())[1])
+    if (await getChainId() == 31337) {
```
Let's not have an "if local" here. Let's have the local and tesnet tests try to be the same whenever possible.

pappas999 replied:
Price feeds is hard to have the same uniform test for both local and public networks because of the way the deployments are done. Local uses MockAggregator with some static value as the price, then we compare the deployed price feed price to that number, whereas public network uses the already deployed aggregators, and just checks for a  >0 value, which means that it was deployed properly and is looking at the aggregator specified in the constructor (although we don't really know if the number is correct)

The public network test can't compare the price feed value to the aggregator contract like the local one (as far as I'm aware), but the local one can just check for a >0 value as opposed to comparing the price feed latest price to the mock aggregator latestRoundData, so for now I'll go with that, but let me know if you have any other suggestions

PatrickAlphaC commented on test/RandomNumberConsumer_test.js:
```diff
@@ -1,29 +1,55 @@
-    const tx_receipt = await transaction.wait()
-    const requestId = tx_receipt.events[2].topics[0]
+    // If local chain, do tests using mocks
+    if (await getChainId() == 31337) {
```
same comment here about having two massive code blocks that are really only have one or two lines different.

PatrickAlphaC reviewed (changes requested):
Please see comments

pappas999 commented:
removed autoFund action outside of the helper function as suggested

also tried to clean up the tests a bit to make them a bit more uniform regardless of network used. Let me know any other feedback

@PatrickAlphaC

thodges-gh commented on contracts/APIConsumer.sol:
```diff
@@ -69,6 +69,17 @@ contract APIConsumer is ChainlinkClient {
+    /**
+     * Withdraw LINK from this contract
+     * 
+     * NOTE: DO NOT USE THIS IN PRODUCTION AS IT CAN BE CALLED BY ANY ADDRESS.
+     * THIS IS PURELY FOR EXAMPLE PURPOSES ONLY.
```
I think this is bad enough practice that we should just make the contract implement Ownable and put an `onlyOwner` modifier on functions.

thodges-gh commented on hardhat.config.js:
```diff
@@ -87,4 +88,4 @@ module.exports = {
         timeout: 100000
     }
-}
+}
```
Modify your editor to include ending newlines.

thodges-gh commented on test/PriceConsumerV3_test.js:
```diff
@@ -1,23 +1,29 @@
 
   beforeEach(async () => {
-    await deployments.fixture(['mocks', 'feed'])
+    // We ONLY deploy the fixtures on a local test
+    if (await getChainId() == 31337) {
+      await deployments.fixture(['mocks', 'feed'])
+    }
```
You should only be running unit tests on local networks.

PatrickAlphaC replied:
Should we have a separate section for running integration tests? I 100% will want to test on a testnet after running my tests locally before.

thodges-gh commented on test/RandomNumberConsumer_test.js:
```diff
@@ -1,29 +1,48 @@
+  
   let randomNumberConsumer, vrfCoordinatorMock, seed
-
+  
```
```suggestion

```

thodges-gh replied:
Also update your editor to display trailing whitespace.

thodges-gh commented on deploy/02_Deploy_APIConsumer.js:
```diff
@@ -23,18 +25,23 @@ module.exports = async ({
-  log("Then run API Consumer contract with following command:")
-  log("npx hardhat request-data --contract " + apiConsumer.address + " --network " + networkConfig[chainId]['name'])
+  //try to auto fund contract with LINK
+  const autoFund = await autoFundCheck(apiConsumer.address, networkName, linkTokenAddress, additionalMessage);
```
This whole autoFund thing confuses me. Why not just fund it if `apiConsumer.newlyDeployed`?

pappas999 replied:
what if you're deploying to mainnet? Or what if there's no LINK in the wallet? The auto fund is doing what you just suggested, except its configurable so it can be turned on/off per network, and it has some smarts to check if there's LINK in the wallet as well

pappas999 commented:
Incorporated feedback:

- removed comments in contracts about not using in PROD, and implemented ownable modifiers
- removed auto-fund parameter, created a new 'setup' deploy script to run after all the previous deployments, this now handles the auto-fund, and is the place where any future 'tasks' can be put 
- provided the ability to deploy contracts without running the setup-contracts script, updated instructions in README
- split the tests up into unit & integration, put in commands in package.json to run each of them separately with a different command

PatrickAlphaC reviewed (approved):
LGTM

thodges-gh commented on deploy/04_Setup_Contracts.js:
```diff
@@ -0,0 +1,56 @@
+  log("----------------------------------------------------")
+  
+}
+module.exports.tags = ['all']
```
File needs ending newline.

thodges-gh commented on deploy/04_Setup_Contracts.js:
```diff
@@ -0,0 +1,56 @@
+  const APIConsumer = await deployments.get('APIConsumer')
+  let apiConsumer = await ethers.getContractAt('APIConsumer', APIConsumer.address)  
+
+  let autoFund = await autoFundCheck(apiConsumer.address, networkName, linkTokenAddress, additionalMessage)
+  if (autoFund == true) {
```
These can be refactored to
```
if (await autoFundCheck(apiConsumer.address, networkName, linkTokenAddress, additionalMessage)) {
```
instead of having to declare `autoFund` and set its value again later.

thodges-gh commented on deploy/04_Setup_Contracts.js:
```diff
@@ -0,0 +1,56 @@
+  const RandomNumberConsumer = await deployments.get('RandomNumberConsumer')
+  let randomNumberConsumer = await ethers.getContractAt('RandomNumberConsumer', RandomNumberConsumer.address)
+
+  autoFund = await autoFundCheck(randomNumberConsumer.address, networkName, linkTokenAddress, additionalMessage)
+  if (autoFund == true) {
```
```
if (await autoFundCheck(randomNumberConsumer.address, networkName, linkTokenAddress, additionalMessage)) {
```

thodges-gh commented on deploy/04_Setup_Contracts.js:
```diff
@@ -0,0 +1,56 @@
+
+  //Try Auto-fund APIConsumer contract with LINK
+  const APIConsumer = await deployments.get('APIConsumer')
+  let apiConsumer = await ethers.getContractAt('APIConsumer', APIConsumer.address)  
```
Should be `const`.

thodges-gh commented on helper-hardhat-config.js:
```diff
@@ -50,8 +56,31 @@ const getNetworkIdFromName = async (networkIdName) => {
+    const signer = accounts[0]
+    const LinkToken = await ethers.getContractFactory("LinkToken")
+    const linkTokenContract = new ethers.Contract(linkTokenAddress, LinkToken.interface, signer)
+    let balanceHex = await linkTokenContract.balanceOf(signer.address);
+    let balance = await web3.utils.toBN(balanceHex._hex).toString();   
```
These should be `const`.

thodges-gh commented on deploy/03_Deploy_RandomNumberConsumer.js:
```diff
@@ -1,11 +1,11 @@
-let { networkConfig } = require('../helper-hardhat-config')
-
+let { networkConfig, autoFundCheck } = require('../helper-hardhat-config')
```
`autoFundCheck` is not used in this file.

thodges-gh commented on deploy/02_Deploy_APIConsumer.js:
```diff
@@ -1,4 +1,4 @@
-let { networkConfig } = require('../helper-hardhat-config')
+let { networkConfig, autoFundCheck } = require('../helper-hardhat-config')
```
`autoFundCheck` is not used in this file.

thodges-gh commented on deploy/02_Deploy_APIConsumer.js:
```diff
@@ -10,6 +10,8 @@ module.exports = async ({
   let oracle
   let additionalMessage = ""
+  //set log level to ignore non errors
+  ethers.utils.Logger.setLogLevel(ethers.utils.Logger.levels.ERROR);
```
Rogue `;`

thodges-gh commented on deploy/04_Setup_Contracts.js:
```diff
@@ -0,0 +1,56 @@
+const { networkConfig, autoFundCheck } = require('../helper-hardhat-config')
+const {ethers, getNamedAccounts} = require('hardhat');
```
Rogue `;`

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+    .addOptionalParam("linkaddress", "Set the LINK token address")
+    .setAction(async (taskArgs) => {
+        const contractAddr = taskArgs.contract
+        let networkId = await getNetworkIdFromName(network.name)
```
Should be `const`.

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+let { networkConfig, getNetworkIdFromName } = require('../helper-hardhat-config')
```
Should be `const`.

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+        const signer = accounts[0]
+
+        //First, lets see if there is any LINK to withdraw
+        let linkTokenAddress = networkConfig[networkId]['linkToken'] || taskArgs.linkaddress
```
Should be `const`.

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+        let linkTokenAddress = networkConfig[networkId]['linkToken'] || taskArgs.linkaddress
+        const LinkToken = await ethers.getContractFactory("LinkToken")
+        const linkTokenContract = new ethers.Contract(linkTokenAddress, LinkToken.interface, signer)
+        let balanceHex = await linkTokenContract.balanceOf(contractAddr);
+        let balance = await web3.utils.toBN(balanceHex._hex).toString(); 
```
Should be `const`.

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+        let linkTokenAddress = networkConfig[networkId]['linkToken'] || taskArgs.linkaddress
+        const LinkToken = await ethers.getContractFactory("LinkToken")
+        const linkTokenContract = new ethers.Contract(linkTokenAddress, LinkToken.interface, signer)
+        let balanceHex = await linkTokenContract.balanceOf(contractAddr);
+        let balance = await web3.utils.toBN(balanceHex._hex).toString(); 
+        console.log('LINK balance of contract: ' + contractAddr + " is " + balance / Math.pow(10,18));
```
Rogue `;`

thodges-gh commented on tasks/withdraw-link.js:
```diff
@@ -0,0 +1,38 @@
+
+            //Create connection to Consumer Contract and call the withdraw function
+            const ConsumerContract = new ethers.Contract(contractAddr, RandomNumberConsumer.interface, signer)
+            let result = await ConsumerContract.withdrawLink().then(function (transaction) {
```
Should be `const`.

thodges-gh commented on test/integration/APIConsumer_int_test.js:
```diff
@@ -0,0 +1,28 @@
+const chai = require('chai')
+const {expect} = require('chai')
+const BN = require('bn.js');
+chai.use(require('chai-bn')(BN));
```
Rogue `;`

thodges-gh commented on test/integration/APIConsumer_int_test.js:
```diff
@@ -0,0 +1,28 @@
+    const requestId = tx_receipt.events[0].topics[1]
+    
+    //wait 30 secs for oracle to callback
+    await new Promise(resolve => setTimeout(resolve, 30000));
```
Rogue `;`

thodges-gh commented on test/integration/RandomNumberConsumer_int_test.js:
```diff
@@ -0,0 +1,28 @@
+const chai = require('chai')
+const {expect} = require('chai')
+const BN = require('bn.js');
+chai.use(require('chai-bn')(BN));
```
Rogue `;`

thodges-gh commented on test/integration/RandomNumberConsumer_int_test.js:
```diff
@@ -0,0 +1,28 @@
+    const requestId = tx_receipt.events[2].topics[1]
+    
+    //wait 30 secs for oracle to callback
+    await new Promise(resolve => setTimeout(resolve, 30000));
```
Rogue `;`

thodges-gh reviewed (changes requested):
Lots of style issues.

pappas999 commented:
- removed .env (to avoid it being accidently checked in)
- Removed extra trailing whitespaces and rogue ; characters
- removed variables not used
- cleaned up ; characters
- changed let to const where required
- updated README

pappas999 commented:
@PatrickAlphaC @thodges-gh I believe I've addressed all the comments, and also incorporated @PatrickAlphaC latest VRF changes as well.

PatrickAlphaC commented:
@pappas999 can you resolve the conversations where changes are requested?

@thodges-gh it looks like the changes have been implemented, can we merge?

PatrickAlphaC commented:
Actually, not yet.

PatrickAlphaC commented:
Harry please see my video on fixing the merge conflicts - I dm'd you

PatrickAlphaC commented:
FYI “bad” refers to my original review being bad
